sktime/forecasting/compose/_skforecast_reduce.py

```diff
@@ -0,0 +1,469 @@
+
+    """
+
+    _tags = {
+        "authors": ["Abhay-Lejith", "yarnabrina"],
+        "maintainers": ["yarnabrina"],
+        "y_inner_mtype": "pd.Series",
+        "X_inner_mtype": "pd.DataFrame",
+        "requires-fh-in-fit": False,
+        "handles-missing-data": False,
+        "capability:insample": False,
+        "capability:pred_int": True,
+        "capability:pred_int:insample": False,
+        "python_version": ">=3.8,<3.13",
+        "python_dependencies": ["skforecast>=0.12.1"],
+    }
+
+    def __init__(
+        self: "SkforecastAutoreg",
+        regressor: object,
+        lags: Union[int, np.ndarray, list],
+        transformer_y: Optional[object] = None,
+        transformer_exog: Optional[object] = None,
+        weight_func: Optional[Callable] = None,
+        differentiation: Optional[int] = None,
+        fit_kwargs: Optional[dict] = None,
+        binner_kwargs: Optional[dict] = None,
+        forecaster_id: Optional[Union[str, int]] = None,
+    ) -> None:
+        self.regressor = regressor
+        self.lags = lags
+        self.transformer_y = transformer_y
+        self.transformer_exog = transformer_exog
+        self.weight_func = weight_func
+        self.differentiation = differentiation
+        self.fit_kwargs = fit_kwargs
+        self.binner_kwargs = binner_kwargs
+        self.forecaster_id = forecaster_id
+
+        super().__init__()
+
+        self._forecaster = None
+        self._transformer_y = None
+        self._transformer_exog = None
+
+        self._clone_estimators()
+
+    def _clone_estimators(self: "SkforecastAutoreg"):
+        """Clone the regressor and transformers."""
+        from sklearn.base import clone
+
+        self._regressor = clone(self.regressor)
+
+        if self.transformer_y:
+            self._transformer_y = clone(self.transformer_y)
+
+        if self.transformer_exog:
+            self._transformer_exog = clone(self.transformer_exog)
+
+    def _create_forecaster(self: "SkforecastAutoreg"):
+        """Create ``skforecast.ForecasterAutoreg.ForecasterAutoreg`` model."""
+        from skforecast.ForecasterAutoreg import ForecasterAutoreg
+
+        return ForecasterAutoreg(
+            self._regressor,
+            self.lags,
+            transformer_y=self._transformer_y,
+            transformer_exog=self._transformer_exog,
+            weight_func=self.weight_func,
+            differentiation=self.differentiation,
+            fit_kwargs=self.fit_kwargs,
+            binner_kwargs=self.binner_kwargs,
+            forecaster_id=self.forecaster_id,
+        )
+
+    @staticmethod
+    def _coerce_column_names(X: Optional[pd.DataFrame]):
+        if X is None:
+            return None
+
+        return X.rename(columns=lambda column_name: str(column_name))
+
+    @staticmethod
+    def _coerce_int_to_range_index(df):
+        new_df = df.copy(deep=True)
+        if new_df is not None:
+            new_index = pd.RangeIndex(new_df.index[0], new_df.index[-1] + 1)
```

Comment:
It will fail in skforecast. Integer index is not supported in skforecast.
Here it will fail saying could not coerce to RangeIndex.

Reply:
Sorry, trying to understand a bit better.

Looking at this (https://skforecast.org/0.12.1/user_guides/input-data.html), it seems non-DatetimeIndex will be handled by skforecast directly. Are you saying that is not happening?

For example, this works for me albeit with warnings. With this PR's code, will the sktime wrapper work?

```python
import numpy
import pandas
from sklearn.linear_model import LinearRegression
from skforecast.ForecasterAutoreg import ForecasterAutoreg

random_generator = numpy.random.default_rng(seed=0)

sample_data = {
    "target_column": random_generator.standard_normal(size=10)
}
sample_index = pandas.Index(random_generator.permutation(10))
sample_dataset = pandas.DataFrame(data=sample_data, index=sample_index)

regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(sample_dataset["target_column"])  # works
print(forecaster.fitted)  # True

forecaster.predict(5)  # works
```

Reply:
That's weird. In the earlier state of this PR, i.e #5447 , it would fail at the below point.
```
index = Index([49, 50, 51], dtype='int64'), steps = 1

    def expand_index(
        index: Union[pd.Index, None],
        steps: int
    ) -> pd.Index:
        """
        Create a new index of length `steps` starting at the end of the index.
    
        Parameters
        ----------
        index : pandas Index, None
            Original index.
        steps : int
            Number of steps to expand.
    
        Returns
        -------
        new_index : pandas Index
            New index.
    
        """
    
        if isinstance(index, pd.Index):
    
            if isinstance(index, pd.DatetimeIndex):
                new_index = pd.date_range(
                                start   = index[-1] + index.freq,
                                periods = steps,
                                freq    = index.freq
                            )
            elif isinstance(index, pd.RangeIndex):
                new_index = pd.RangeIndex(
                                start = index[-1] + 1,
                                stop  = index[-1] + 1 + steps
                            )
        else:
            new_index = pd.RangeIndex(
                            start = 0,
                            stop  = steps
                        )
    
>       return new_index
E       UnboundLocalError: local variable 'new_index' referenced before assignment
```
So integer and period indexes were not working. So I made the function to convert the index types.

Reply:
>Looking at this (https://skforecast.org/0.12.1/user_guides/input-data.html), it seems non-DatetimeIndex will be handled by skforecast directly. Are you saying that is not happening?

Not when we pass X. skforecast doesn't seem to convert the index type of X.  It fails with error saying index must be range or datetime.
Try this snippet passing X:
```python
import numpy
import pandas
from sklearn.linear_model import LinearRegression
from skforecast.ForecasterAutoreg import ForecasterAutoreg
from sktime.split import temporal_train_test_split

random_generator = numpy.random.default_rng(seed=0)

list=[1,2,3,4,5,6,7,8]
sample_data = list
sample_index = pandas.Index(list)
y = pandas.Series(data=sample_data, index=sample_index, name="y")

X = pandas.DataFrame(data=sample_data ,index=sample_index,columns=["X"])


y_train,y_test,X_train,X_test=temporal_train_test_split(y,X,0.5)
regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(y_train,X_train)  
print(forecaster.fitted) 
forecaster.predict(4,exog=X_test)
```

> With this PR's code, will the sktime wrapper work?

I didn't understand, this PR's code is working right? Otherwise how are all tests passing?

Reply:
I'll try with X tonight after office, will revert.

As of now, looking at the code I think just y with non-sequential index (similar to my snippet above) will fail with your adapter class, even though it works directly with skforecast. Can you please confirm if that is my misunderstanding?

> Otherwise how are all tests passing?

I am not sure if sktime's default testing suite contains test case with non-sequential integers being the index or similar. May be @fkiraly can confirm on that, but we can also add a new test case.

Reply:
> As of now, looking at the code I think just y with non-sequential index (similar to my snippet above) will fail with your adapter class.

Yes, if Integer index is passed and it is non-sequential. It will fail here because it won't be converted to RangeIndex correctly. 
This is a consequence of the index type conversion that I am doing in order to support all index types in exog. I will see if I can figure out a way to convert non-sequential index correctly.
Or, I can make the index conversion only when `X` is passed. So in your example above, since no `X` is passed, I don't convert index type and it will run without error. Only downside with this is that non-sequential index in `X` and `y` will not work. But this does not work in skforecast either.

> I am not sure if sktime's default testing suite contains test case with non-sequential integers being the index or similar.

If I'm not wrong, I think all fixtures have sequential indices.

Reply:
@yarnabrina I checked with some other models, non-sequential indices do not work in sktime. I was getting errors saying that index must be monotonically increasing.

Reply:
> Not when we pass X. It fails with error saying index must be range or datetime.

Yes, I am facing that only for prediction, training seems to happen fine. Interestingly, even with range index I am facing the same error.

```python
import numpy
import pandas
from skforecast.ForecasterAutoreg import ForecasterAutoreg
from sklearn.linear_model import LinearRegression

random_generator = numpy.random.default_rng(seed=0)

sample_data = {
    "target_column": random_generator.standard_normal(size=10),
    "numeric_column": random_generator.standard_normal(size=10),
}
sample_index = pandas.Index(random_generator.permutation(10))
sample_dataset = pandas.DataFrame(data=sample_data, index=sample_index)

regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(sample_dataset.iloc[:7, 0], exog=sample_dataset.iloc[:7, 1])  # works
print(forecaster.fitted)  # True

forecaster.predict(3, exog=sample_dataset.iloc[7:, 1])  # fails

hacky_future_exog_dataset = sample_dataset.iloc[7:, 1]
hacky_future_exog_dataset.index = pandas.RangeIndex(start=7, stop=10, step=1)

forecaster.predict(3, exog=hacky_future_exog_dataset)  # fails (why?)
```

Regarding how should we fail, I am thinking if we can use some tags for this purpose. A tag can be used to show if a particular estimator has the capability to forecast in presence of gaps in the data, and for all such cases we can fail with a common error. @fkiraly / @benHeid any thoughts?

As of now, the only suggestion I have is to allow step to be something other than 1, everything else seems fine. The way you are creating `new_index` implies `step=1` always, can we allow that as the gap between positions 0 and 1?

Reply:
Yes, we are assuming step=1 here. I guess we can also make step= index[1]-index[0] but this would assume that the indexes are equally spaced apart. 
If the integer index isn't spaced equally apart, I don't see any solution as it cannot be converted to any supported index type.

Reply:
Right, and we'll keep it as it is. So, for now, I suggest you make two changes: allow constant step (but not necessarily 1) and in the error message inform whether it's failing for `y` or `X`.

We did a similar type of thing in `neuralforecast` adapter earlier, if I remember correctly. Not on laptop at the moment to check, but if possible try to use similar error message if possible.